Building an API pt. 3 - lesson 2: "Models Setup" updates (#147)
Ada-Developers-Academy/core-unit-2: api-3-database-models-read/models-setup.md

Reviewer: Yep. That's it.

## Removing Hardcoded Data

In a previous lesson we used hardcoded data stored in the variable `books` to learn how to write routes to _read_ all books, and _read_ one book. This was great for learning, but we were limited to _GET_ routes, as we could not persist any changes to our data given that it was hardcoded.

In this lesson, we will learn how to connect our Flask app to a database. This will allow us to not only write routes to _read_ our data, but also _create_, _update_, and _delete_ it!

> Before we go any further, let's comment or remove our hardcoded `books` data and our two routes from the `app`.

Reviewer: I looked through some other lessons to check for consistency, and it looks like with the rest of the lessons that have a collapsible section after the branch table have them abut, so that should be left as is. The same for any that abut a Piece of Code table. The way everything fits together in those cases make it look like an intentional UI eleemnt.

But for cases of a summary detail following a paragraph, adding the line break helps the paragraph feel less cramped.

> 

Reviewer: Looks good. For the heading, how about

```
## The `psql` Prompt Provides Input Feedback
```

## Connecting the Database and Flask

There are a couple steps to creating the database connection. We will:
1. create objects that represent our database and migration tools
2. register these objects with our application and tell Flask where find our new database in `create_app`

### Set Up the Database and Migrations

To set up our connection with SQLAlchemy, we need to create an instance of SQLAlchemy and give it information on how we want to create our database backed models. We do this by passing the SQLAlchemy constructor a class name to use when generating our model classes.

> That might sound like a lot of work, but SQLAlchemy provides a class named `DeclarativeBase` that provides all of the operations our database backed models will need. We will make our own `Base` model class that is a subclass of `DeclarativeBase`. 

Reviewer: I agree with swapping "that" to "which", since it's a non-restrictive clause. In that case, a comma is also recommended. In the second sentence, "that will be" has closer verb agreement with "we will make".
```suggestion
That might sound like a lot of work, but SQLAlchemy provides a class named `DeclarativeBase`, which provides all of the operations our database backed models will need. We will make our own `Base` model class that will be a subclass of `DeclarativeBase`. 
```

While we will not dive further into customizing our `Base` class in this series, creating our own `Base` class means that we could further tailor the behavior of our model classes, adding operations that we want available to all models. 

Following the pattern of creating a file for every model, we can create a file for `Base` in our `models` folder:

```bash
$ touch app/models/base.py
```

Let's take a look at the code for our `Base` class before copying it over to our new file:

```py
from sqlalchemy.orm import DeclarativeBase

class Base(DeclarativeBase):
    pass
```

Our code here is short, our only actions are to 
1. import the parent class `DeclarativeBase`
2. create an empty subclass `Base` that we can give to our SQLAlchemy constructor.

With our `Base` model in hand, we can create a new file in the `app` folder `db.py` which will manage creating our database and migration objects:

```bash
$ touch app/db.py
```

> We can read through the contents of that file:

Reviewer: That works, but I'm still bothered by "we can".

```suggestion
Let's look over the contents for `db.py` before copying it to our local file:
```

3. Last, we make an instance of Migrate. We will not directly interact with this object much, but this object will be used by the application to update our database tables when we make changes to our model class's attributes. 

At this point, our current project structure likely looks similar to this:

```
.
├── app
│   ├── models
│   │   ├── __init__.py
│   │   ├── base.py
│   │   └── book.py
│   ├── routes
│   │   ├── __init__.py
│   │   └── book_routes.py
│   ├── __init__.py
│   └── db.py
├── README.md
└── requirements.txt
```

### Configure the Database in `app/__init__.py`

To finish configuring the database, we will make 2 changes to `create_app`:
- Supply a _connection string_ to tell Flask where to find our new database
- import the `db` and `migrate` objects we created and call `init_app` on each of them

When we need to tell Flask where to find a database, we do this by providing a _connection string_ that identifies where it is, and how to connect to it.

A connection string resembles the following:

```
postgresql+psycopg2://postgres:postgres@localhost:5432/REPLACE_THIS_LAST_PART_WITH_DB_NAME
```

> 

Author: Agreed, I can revise that paragraph into a table and it should be easier to take in.

## The `psql` Prompt Provides Input Feedback

If a `psql` command isn't responding like we expect, we should check our `psql` prompt to make sure our commands are being interpreted correctly. As a reminder:
- an `=` between the database name and the `#` in the prompt indicates `psql` is waiting for the start of a new command, but if we see `-`, that means it's continuing from the previous line (likely because we forgot a `;`).
- SQL query commands will not be run unless they end in a `;`, but admin commands (such as `\l` and `\q`) don't require a terminal `;`.

### !end-callout

## Connecting the Database and Flask

There are a couple steps to creating the database connection. We will:
1. initialize objects to represent our database, and migration tools
2. register these objects with our application and tell Flask where to find our new database in `create_app`

### Set Up the Database and Migrations

To set up our connection with SQLAlchemy, we need to create an instance of `SQLAlchemy` and give it information on how we want to create our database backed models. We do this by passing the SQLAlchemy constructor a class name to use when generating our model classes.

> That might sound like a lot of work, but SQLAlchemy provides a class named `DeclarativeBase` which provides all of the operations our database backed models will need. We will make our own `Base` model class that will be a subclass of `DeclarativeBase`. 

Reviewer: I'd still include this comma for the non-restrictive clause
```suggestion
That might sound like a lot of work, but SQLAlchemy provides a class named `DeclarativeBase`, which provides all of the operations our database backed models will need. We will make our own `Base` model class that will be a subclass of `DeclarativeBase`. 
```

While we will not dive further into customizing our `Base` class in this series, creating our own `Base` class means that we could further tailor the behavior of our model classes, adding operations that we want available to all models. 

Following the pattern of creating a file for every model, we can create a file for `Base` in our `models` folder:

```bash
$ touch app/models/base.py
```

Let's take a look at the code for our `Base` class before copying it over to our new file:

```py
from sqlalchemy.orm import DeclarativeBase

class Base(DeclarativeBase):
    pass
```

Our code here is short, with our only actions being to 
1. import the parent class `DeclarativeBase`, and
2. create an empty subclass `Base` that we can give to our SQLAlchemy constructor.

> With our `Base` model in hand, we can create a new file in the `app` folder `db.py` which will manage creating our database and migration objects:

Reviewer: non-restrictive clause comma
```suggestion
With our `Base` model in hand, we can create a new file in the `app` folder `db.py`, which will manage creating our database and migration objects:
```

```bash
$ touch app/db.py
```

Let's look over the contents for `db.py` before copying it to our local file:

```py
from flask_sqlalchemy import SQLAlchemy
from flask_migrate import Migrate
from .models.base import Base

db = SQLAlchemy(model_class=Base)
migrate = Migrate()
```

Then we can update our `db.py` to contain the same code.

The file above does a few things:

1. First we import `SQLAlchemy`, `Migrate`, and our newly created `Base` class

> 2. Next, we create an instance of SQLAlchemy that we will call `db` and pass it our `Base` class as the constructor argument. We will use the `db` object when we need to interact with the database to perform operations like creating or updating records.

Reviewer: non-restrictive clause
```suggestion
2. Next, we create an instance of SQLAlchemy, which we will call `db`, and pass it our `Base` class as the constructor argument. We will use the `db` object when we need to interact with the database to perform operations like creating or updating records.
```

At this point, our project structure should look similar to this:

```
.
├── app
│   ├── models
│   │   ├── __init__.py
│   │   ├── base.py
│   │   └── book.py
│   ├── routes
│   │   ├── __init__.py
│   │   └── book_routes.py
│   ├── __init__.py
│   └── db.py
├── README.md
└── requirements.txt
```

### Configure the Database in `app/__init__.py`

To finish configuring the database, we will make 2 changes to `create_app`:
- Supply a _connection string_ to tell Flask where to find our new database
- import the `db` and `migrate` objects we created and use their `init_app` methods to register them with our `app`

When we need to tell Flask where to find a database, we do this by providing a _connection string_ that identifies where it is, and how to connect to it.

The required format of a connection string can vary depending on the library being used to communicate with the database. The packages we are using (SQLAlchemy and psycopg2) structure their connection strings to look like a URL. Consider the example connection string shown here:

> ```

Reviewer: undecorated, there's some weird coloration going on
```suggestion
```text
```

postgresql+psycopg2://postgres:postgres@localhost:5432/REPLACE_THIS_LAST_PART_WITH_DB_NAME
```

| <div style="min-width:250px;"> Piece of Code </div> | Notes|
| --------------------------------------------------- | ---- |

> | `postgresql+psycopg2` | This tells Flask to connect to our database using the `psycopg2` package we installed from our `requirements.txt`. |

Reviewer: Thanks for including the table! How about chunking them up like this (and the following)
```suggestion
| `postgresql+psycopg2` | Contains information about the protocols to use for connecting to the database. This tells Flask to connect to our `postgresql` database using the `psycopg2` package we installed from our `requirements.txt`. |
```

Author: Definitely more complete information, thanks for including the suggestions!

> | `://postgres:postgres` | We connect to the database using the `postgres` protocol and the user `postgres` |

Reviewer: ```suggestion
| `://` | Required URL syntax to separate the protocol part from the user or host part. |
| `postgres:postgres` | Contains information about the user credentials to use to connect to a resource, here, the database. We connect to the database using the `postgres` user (the part before the `:`), using the password `postgres` (the part after the `:`). We never actually set a database password, but when connecting to a local PostgreSQL service, the supplied password value is ignored anyway and could be omitted. It is included here for completeness. |
```

> | `@localhost:5432` | We are connecting to the local machine (`localhost`) running at port `5432` |

Reviewer: ```suggestion
| `@` | Required URL syntax to separate the user part from the host part. |
| `localhost:5432` | Contains information about the host (the machine "hosting" the database) location. We are connecting to the local machine (`localhost`) running at port `5432` |
```

> | `REPLACE_THIS_LAST_PART_WITH_DB_NAME` | This text should be replaced with the name of the database we're connecting to. For the hello-books-api project, we will replace it with our development database's name, `hello_books_development`. |

Reviewer: ```suggestion
| `/` | Required URL syntax to separate the host part from the path part. |
| `REPLACE_THIS_LAST_PART_WITH_DB_NAME` | Normally contains path information about the resource being accessed. This text should be replaced with the name of the database we're connecting to. For the hello-books-api project, we will replace it with our development database's name, `hello_books_development`. |
```

### Our Models Are Classes That Inherit From `db.Model`

SQLAlchemy provides a pattern for creating a class for our model that will:

1. Connect the model seamlessly with our SQL database, mapping attributes to table columns
2. Give our Flask code the ability to work with `Book` instances, taking advantage of OOP
3. Enable us to access instances of `Book`, where each instance corresponds to a row in our database

```python
from sqlalchemy.orm import Mapped, mapped_column
from ..db import db

class Book(db.Model):
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    title: Mapped[str]
    description: Mapped[str]
```

### !callout-info

## Be Aware of Changes to SQLAlchemy When Looking at Older Examples

SQLAlchemy has changed over the years, both in syntax and in meaning. The current syntax borrows from Python's type hinting system to make the column declarations more concise, as well as to make it easier for tools to check our code for errors. Additionally, some behaviors that used to be the default must now be set explicitly.

<br />

> Recall that SQL permits us to define columns that allow NULL values. In the past, SQLAlchemy would default to allowing NULL values for all columns unless otherwise marked. In the current version, columns using `Mapped` types as above, that are not explicitly marked `Optional`, are considered non-nullable; they are required attributes when creating a new instance of a model.

Reviewer: shifts the aside to avoid splitting the restrictive clause
```suggestion
Recall that SQL permits us to define columns that allow NULL values. In the past, SQLAlchemy would default to allowing NULL values for all columns unless otherwise marked. In the current version, columns using `Mapped` types that are not explicitly marked `Optional`, as above, are considered non-nullable; they are required attributes when creating a new instance of a model.
```

> ### Making the `Book` Model Visible to the `app` in `__init__.py`

Reviewer: I'm good with going all the way to Alembic. In the past, we've sort of downplayed its involvement, but at the end of the day, all those `flask db` commands are Alembic commands, and those are the commands that need the models to be visible prior to explicitly using them anywhere else.

(venv) $ flask db init
```

Example Output:

```bash
(venv) $ flask db init
  Creating directory '/Users/user_name/Documents/hello-books-api/migrations' ...  done
  Creating directory '/Users/user_name/Documents/hello-books-api/migrations/versions' ...  done
  Generating /Users/user_name/Documents/hello-books-api/migrations/script.py.mako ...  done
  Generating /Users/user_name/Documents/hello-books-api/migrations/env.py ...  done
  Generating /Users/user_name/Documents/curriculum_changes/hello-books-api/migrations/README ...  done
  Generating /Users/user_name/Documents/curriculum_changes/hello-books-api/migrations/alembic.ini ...  done
  Please edit configuration/connection/logging settings in
  '/Users/user_name/Documents/hello-books-api/migrations/alembic.ini' before proceeding.
```

> Although the command `flask db init` sounds like it should initialize our database, in reality, it doesn't look at the database at all! This command creates our project's migrations folder, which Alembic will use to store its generated migration files.

Reviewer: ```suggestion
Although the command `flask db init` sounds like it should initialize our database, in reality, it doesn't look at the database at all! This command creates our project's `migrations` folder, which Alembic will use to store its generated migration files.
```

When the migrate command is successful, we should see something similar to the last 2 lines:
```bash
INFO  [alembic.autogenerate.compare] Detected added table 'book'
  Generating /Users/user_name/Documents/hello-books-api/migrations/versions/bf0d3269e52e_adds_book_model.py ...  done
```
where an `INFO` line calls out the detected change, followed by a final line describing where the new migration file was created.

If we haven't made any changes to our models that require the database to be modified, or if some other issue is preventing the migration tool from seeing changes, we will often see an `INFO` message at the end of the output stating  `"No changes in schema detected"`:

Example Error Output:

```bash
(venv) $ flask db migrate -m "adds Book model"
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.ddl.postgresql] Detected sequence named 'book_id_seq' as owned by integer column 'book(id)', assuming SERIAL and omitting
INFO  [alembic.env] No changes in schema detected.
```

> If no change is detected, we should reflect on whether any recent model changes would require the database to be updated. For instance, if we add a helper method to a model class, only our code is affected, it doesn't change the data stored in the database. In that case, generating a migration isn't necessary, and if we try to do so, no changes will be detected. 

Reviewer: separates the two clauses. could use ; instead
```suggestion
If no change is detected, we should reflect on whether any recent model changes would require the database to be updated. For instance, if we add a helper method to a model class, only our code is affected. It doesn't change the data stored in the database. In that case, generating a migration isn't necessary, and if we try to do so, no changes will be detected. 
```

(venv) $ flask db init
```

Example Output:

```bash
(venv) $ flask db init
  Creating directory '/Users/user_name/Documents/hello-books-api/migrations' ...  done
  Creating directory '/Users/user_name/Documents/hello-books-api/migrations/versions' ...  done
  Generating /Users/user_name/Documents/hello-books-api/migrations/script.py.mako ...  done
  Generating /Users/user_name/Documents/hello-books-api/migrations/env.py ...  done
  Generating /Users/user_name/Documents/curriculum_changes/hello-books-api/migrations/README ...  done
  Generating /Users/user_name/Documents/curriculum_changes/hello-books-api/migrations/alembic.ini ...  done
  Please edit configuration/connection/logging settings in
  '/Users/user_name/Documents/hello-books-api/migrations/alembic.ini' before proceeding.
```

The final message in the output `"Please edit configuration..."` is just a recommendation. We will not be discussing logging data about our API as part of this series, so we do not need to take any action with the listed file. Once we see this message we are ready to keep moving forward!

### Generate Migrations After Each Model Change

> We can generate database migrations with the following command. This command should be run every time there's a change to a file in the `models` folder.

Reviewer: Works for me!

When the migrate command is successful, we should see something similar to the last 2 lines:
```bash
INFO  [alembic.autogenerate.compare] Detected added table 'book'
  Generating /Users/user_name/Documents/hello-books-api/migrations/versions/bf0d3269e52e_adds_book_model.py ...  done
```
where an `INFO` line calls out the detected change, followed by a final line describing where the new migration file is being created.

If a change is not seen in our model classes, or some other issue is preventing the migration tool from seeing changes, we will often see an `INFO` message at the end of the output stating  `"No changes in schema detected"`:

Example Error Output:

```bash
(venv) $ flask db migrate -m "adds Book model"
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.ddl.postgresql] Detected sequence named 'book_id_seq' as owned by integer column 'book(id)', assuming SERIAL and omitting
INFO  [alembic.env] No changes in schema detected.
```

> 

Author: What do you think about this updated wording & callout?

```suggestion
If no change is detected, we should reflect on whether any recent model changes would require the database to be updated. For instance, if we add a helper method to a model class, only our code is affected, it doesn't change the data stored in the database. In that case, generating a migration isn't necessary, and if we try to do so, no changes will be detected. 

However, if we added a new attribute to our model, that _does_ need to update the database. We need to generate a migration to tell the database to create a new column for the new attribute. If generating a migration in that case detects no changes, we'll need to do more research to determine the issue.
	
### !callout-info

## Database Migration Issues 

Migrations can be a bit finicky to keep in sync, especially when multiple people are working on a project. We need to be careful about coordinating who is creating migrations and ensuring they are shared with the team to keep everyone's development database in sync. 

<br/>

We won't have the opportunity to invest much time in troubleshooting migrations during this series, but a common resolution can be to clear out the database and generate a fresh set of migrations. More details on how to do this are described in the "Resolving Schema Conflicts" reading a bit later on.

### !end-callout
```

> 

Reviewer: Love this, especially mentioning the effect of having multiple contributors. That was something that had crossed my mind that we should be more explicit about at least in roundtables, but seemed like it might be too much to give the consideration it needs what with everything else we're doing. Like it feels like there's enough content to justify a lesson on coordinating migrations on projects with multiple contributors.

But having any kind of mention is better than nothing! 🙏

└── requirements.txt
```

In situations where we need more control over the migration process, we may need to review a generated migration file and modify either the `upgrade` or `downgrade` functions. Pregenerated comments in those functions even encourage us to make adjustments! 

> We probably won't need to make such modifications at this point, but it's a good habit to at least review the migrations before applying them in order to confirm that the generated result matches our expectations. Skimming through the migration files is also good practice for reading code and making assumptions about what it will do based on the context.

Author: I was struggling with the one sentence alone that didn't feel fully thematically linked and made some small adjustments, let me know what you think.

Reviewer: Works for me!

> 

Reviewer: I had made a suggestion to replace the "At the introductory..." passage above with
```md
In situations where we need to have more control over the migration process, we may need to review a generated migration file and modify the code in either the `upgrade` or `downgrade` functions. Pregenerated comments in those functions even encourage us to make adjustments! We probably won't need to make such modifications at this point, but it's good practice to at least review the migrations before applying them in order to confirm that the generated result matches our expectations.

Skimming through the generated migrations is good practice for reading code and making assumptions about what it will do based on the context.
```

## Check for Understanding

<!-- prettier-ignore-start -->
### !challenge
* type: ordering
* id: 304e964c-dcdc-463f-8dcb-3b8843857ecd
* title: Models Setup
##### !question

Arrange the commands below in the correct order to set up a database, create a migration, then apply it.

##### !end-question
##### !answer

1. `psql -U postgres`
2. `CREATE DATABASE hello_books_development;`
3. `flask db init`
4. `flask db migrate -m "Migration Description"`

> 5. `flask db upgrade`

Reviewer: Should we also include `\q` to mark getting out of `psql`. Otherwise, the implication could be that the `flask` commands are run from `psql`.

Reviewer: This could also lead to a hint like "think about which commands are run directly in the database, and which are run from the shell."

We might then include an explanation that list the commands in order, and provides a brief summary of each

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: 459535b6-8a7f-4487-a79e-7ade8fa47884
* title: Models Setup
##### !question

Select all of the options below which could help us troubleshoot if we have an issue connecting to our database

##### !end-question
##### !options

> * Run the app and try to make a request to a route

Reviewer: I'm on the fence with this one. While sometimes a bad connection can result in a crash before the app starts listening (e.g., didn't set the connection string so it ends up None, or maybe a bad protocol is listed leading to plugin load failure), other configurations aren't detected until a request comes in (bad user/pwd/server/port/dbname). So issuing a request and reading the log error can tell you what the issue is. This overlaps with visually inspecting the connection string, but even the case where the app fails to load can be informative, since it can still help us differentiate between a few cases.

So I guess this feels more like a gotcha to me.

Author: Ahh, that's fair and something to avoid. I may just remove this option.

* Check that the database exists locally using `psql`
* Check that the database connection string doesn’t have typos
* Check that `app.config['SQLALCHEMY_TRACK_MODIFICATIONS']` is set to False

> * Check that we are setting `app.config['SQLALCHEMY_DATABASE_URI']` in `__init__.py`

Reviewer: Does the labelled syntax work for checkbox style questions?

Author: I will play around with it and find out, I'm not sure

##### !end-options
##### !answer

* Check that the database exists locally using `psql`
* Check that the database connection string doesn’t have typos
* Check that we are setting `app.config['SQLALCHEMY_DATABASE_URI']` in `__init__.py`

##### !end-answer
##### !explanation

> * Run the app and try to make a request to a route - If the database connection can't be found, often the app will not compile so we cannot use this option.

Reviewer: I'd go with fails to startup rather than compile. But as mentioned above, I would argue this is a still a useful strategy and should be included as a correct answer, since the error message, whether at startup or when the request is sent, can help us determine the issue.

## Check for Understanding

<!-- prettier-ignore-start -->
### !challenge
* type: ordering
* id: 304e964c-dcdc-463f-8dcb-3b8843857ecd
* title: Models Setup
##### !question

Arrange the commands below in the correct order to set up a database, create a migration, then apply it.

##### !end-question
##### !answer

1. `psql -U postgres`
2. `CREATE DATABASE hello_books_development;`
3. `\q`
4. `flask db init`
5. `flask db migrate -m "Migration Description"`
6. `flask db upgrade`

##### !end-answer
##### !hint

Think about which commands are run directly in the database, and which are run from the shell.

##### !end-hint
##### !explanation

1. `psql -U postgres` - start the `psql` command line interface with the user postgres
2. `CREATE DATABASE hello_books_development;` - Create a database named `hello_books_development`

> 3. `\q``- Quit the `psql` command line interface

Reviewer: ```suggestion
3. `\q` - Quit the `psql` command line interface
```

4. `flask db init` - Initialize the migrations folder for our project
5. `flask db migrate -m "Migration Description"` - Create a migration file with the description message "Migration Description"
6. `flask db upgrade` - Apply any new migration files found in the migrations folder

##### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: 459535b6-8a7f-4487-a79e-7ade8fa47884
* title: Models Setup
##### !question

Select all of the options below which could help us troubleshoot if we have an issue connecting to our database.

##### !end-question
##### !options

a| Check that the database exists locally using `psql`
b| Check that the database connection string doesn’t have typos
c| Check that `app.config['SQLALCHEMY_TRACK_MODIFICATIONS']` is set to False
d| Check that we are setting `app.config['SQLALCHEMY_DATABASE_URI']` in `__init__.py`

##### !end-options
##### !answer

> a| Check that the database exists locally using `psql`
> b| Check that the database connection string doesn’t have typos
> d| Check that we are setting `app.config['SQLALCHEMY_DATABASE_URI']` in `__init__.py`

Reviewer: With the labeled options, the solution should be able to be listed as just
```suggestion
a|
b|
d|
```

##### !end-answer
##### !explanation

a| Check that the database exists locally using `psql` - We should check if the database exists and create one if it is not present.

b| Check that the database connection string doesn’t have typos - If we misspell any part of the connection string, we will not be able to connect to the database.

c| Check that app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] is set to False - This setting doesn't affect our ability to reach the database

> d| Check that we are setting app.config['SQLALCHEMY_DATABASE_URI'] in `__init__.py` - We need to give this dictionary key the value of our database connection string for Flask to know where the database is.

Reviewer: In the explanation, the labels aren't treated specially, so a plain bullet will have cleaner formatting.
